collect_app/src/test/java/org/odk/collect/android/external/ExternalDataReaderTest.java

```diff
@@ -0,0 +1,225 @@
+package org.odk.collect.android.external;
+
+import android.database.Cursor;
+import android.database.sqlite.SQLiteDatabase;
+import android.os.Environment;
+
+import org.apache.commons.io.IOUtils;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.odk.collect.android.storage.StorageInitializer;
+import org.odk.collect.android.storage.StoragePathProvider;
+import org.odk.collect.android.storage.StorageSubdirectory;
+import org.odk.collect.android.utilities.CustomSQLiteQueryBuilder;
+import org.odk.collect.android.utilities.FileUtils;
+import org.odk.collect.android.utilities.SQLiteUtils;
+import org.robolectric.RobolectricTestRunner;
+import org.robolectric.shadows.ShadowEnvironment;
+
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.Writer;
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.CoreMatchers.not;
+import static org.hamcrest.CoreMatchers.notNullValue;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.odk.collect.android.external.ExternalDataUtil.COLUMN_DATASET_FILENAME;
+import static org.odk.collect.android.external.ExternalDataUtil.COLUMN_LAST_MODIFIED;
+import static org.odk.collect.android.external.ExternalDataUtil.EXTERNAL_DATA_TABLE_NAME;
+import static org.odk.collect.android.external.ExternalDataUtil.EXTERNAL_METADATA_TABLE_NAME;
+
+@RunWith(RobolectricTestRunner.class)
+public class ExternalDataReaderTest {
+    private static final String SIMPLE_SEARCH_EXTERNAL_CSV_FORM_FILENAME = "simple-search-external-csv.xml";
+    private static final String SIMPLE_SEARCH_EXTERNAL_CSV_NAME = "simple-search-external-csv-fruits";
+    private static final String SIMPLE_SEARCH_EXTERNAL_CSV_FILENAME = "simple-search-external-csv-fruits.csv";
+    private static final String SIMPLE_SEARCH_EXTERNAL_DB_FILENAME = "simple-search-external-csv-fruits.db";
+
+    private static final String SELECT_ALL_DATA_QUERY =  "SELECT * FROM " + EXTERNAL_DATA_TABLE_NAME;
+
+    private static File csvFile;
+    private static File dbFile;
+
+    private static Map<String, File> formDefToCsvMedia;
+
+    @Before
+    public void setUp() throws IOException {
+        ShadowEnvironment.setExternalStorageState(Environment.MEDIA_MOUNTED);
+        new StorageInitializer().createOdkDirsOnStorage();
+
+        File formFile = new File(new StoragePathProvider().getDirPath(StorageSubdirectory.FORMS) + File.separator + SIMPLE_SEARCH_EXTERNAL_CSV_FORM_FILENAME);
```

Comment:
This work here is basically exactly what `CopyFormRule` and the methods in `FormLoadingUtils` do but they're in different sourcesets. Again, maybe a reason to have all tests in one.

Reply:
I'm happy with this for now!